From 11a26da925eebb5cbc1a5dcddf1c5e1f2a62dae2 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Mon, 23 Jun 2014 21:51:54 -0400
Subject: [fs] nfsd: fix bug for readdir of pseudofs

Message-id: <1403560314-32525-173-git-send-email-bfields@redhat.com>
Patchwork-id: 84520
O-Subject: [Patch RHEL7 172/172] NFSD: fix bug for readdir of pseudofs
Bugzilla: 1109733 1111719
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Kinglong Mee <kinglongmee@gmail.com>

Commit 561f0ed498ca (nfsd4: allow large readdirs) introduces a bug
about readdir the root of pseudofs.

Call xdr_truncate_encode() revert encoded name when skipping.

Signed-off-by: Kinglong Mee <kinglongmee@gmail.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: f41c5ad2ff2657978a9712b9ea80cd812a7da2b0
Bugzilla: 1111719
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4xdr.c b/fs/nfsd/nfs4xdr.c
index 2d305a1..83baf2b 100644
--- a/fs/nfsd/nfs4xdr.c
+++ b/fs/nfsd/nfs4xdr.c
@@ -2687,6 +2687,7 @@ nfsd4_encode_dirent(void *ccdv, const char *name, int namlen,
   nfserr = nfserr_toosmall;
   goto fail;
  case nfserr_noent:
+  xdr_truncate_encode(xdr, start_offset);
   goto skip_entry;
  default:
   /*
-- 
1.7.1